packages/cloud_firestore/cloud_firestore_platform_interface/lib/src/method_channel/method_channel_load_bundle_task.dart

```diff
@@ -0,0 +1,52 @@
+import 'dart:async';
+import 'dart:typed_data';
+
+import 'package:cloud_firestore_platform_interface/cloud_firestore_platform_interface.dart';
+
+import 'method_channel_firestore.dart';
+
+import 'utils/exception.dart';
+
+class MethodChannelLoadBundleTask extends LoadBundleTaskPlatform {
+  MethodChannelLoadBundleTask(this._task, this._bundle, this._firestore)
+      : super() {
+    _controller = StreamController<LoadBundleTaskSnapshotPlatform>.broadcast(
+        onCancel: () {
+      nativePlatformStream?.cancel();
+    });
+
+    _task.then((observerId) {
```

Comment:
Using `async*` to do the mapping instead of manually adding listeners & co is likely going to make this a lot more readable and less bug prone

Reply:
Here's an example:

```dart
class Example {
  Example(Future<String> task) {
    Stream<LoadBundleTaskSnapshotPlatform> mapNativeStream() async* {
      final observerId = await task;

      final nativePlatformStream =
          MethodChannelFirebaseFirestore.loadBundleChannel(observerId!)
              .receiveBroadcastStream(
        <String, Object>{'bundle': bundle, 'firestore': firestore},
      ).handleError(
        convertPlatformException,
        test: (err) => err is PlatformException,
      );

      await for (final snapshot in nativePlatformStream) {
        final taskState = convertToTaskState(snapshot['taskState']);

        yield LoadBundleTaskSnapshotPlatform(
          taskState,
          snapshot,
        );

        if (taskState == TaskState.success) {
          // this will close the stream and stop listening to nativePlatformStream
          return;
        }
      }
    }

    stream = mapNativeStream().asBroadcastStream(
      onCancel: (sub) => sub.cancel(),
    );
  }

  late final Stream stream;
}
```